### DISCUSSION TITLE
Use a safer auto-discover default

### DISCUSSION THREAD

**Discussion by ragingdave** [spatie/laravel-settings#56] [Ideas]
Currently settings auto-discover will attempt to discover all classes in the app path. This includes things like a Telescope provider that's only intended to be accessed in a dev/testing environment. This chewed up about an hour or so during a production release. Luckily no downtime as we are using zero downtime deployment methods and this was caught before any damage, but still wasn't caught until prod.

Perhaps a more sane default would be, albeit opinionated, but using app\Settings as a default discover path? I would imagine most apps would warrant a structure such as that, and would prevent being blindsided by this broad auto-discover logic.

**Comment by rubenvanassche** [comment]
Hi @ragingdave,

It is possible to set the paths where classes are auto discovered in settings.php `auto_discover_settings`. How did the the Telescope provider block your site? 

We're not seeing such problems. As for the default of `App/Settings` all our projects do not follow this default since we're using domains to structure our code. So I think we should keep the default as it is.

**Reply by ragingdave** [reply]
That was the ultimate fix of just changing the auto discover path but as far as how the telescope provider blocked our site is as follows. So in dev we always have telescope installed so when the auto discover ran on the app path everything worked fine. This was due to the package's TelescopeApplicationServiceProvider being available as a class in dev as it got installed via composer. In production, where telescope is a dev only dependency, the TelescopeApplicationServiceProvider isn't available as a class, and the discover will cryptically fail right as the package:discover runs after composer install.

Perhaps if the default is going to stay the same, maybe just highlight with a bit more emphasis on the importance of updating that auto discover path during initial integration with this package.

**Reply by rubenvanassche** [reply]
Can we automatically remove this service provider? Please correct me here:

- you have a TelescopeServiceProvider which extends from a BaseTelescopeServiceProvider
- this BaseTelescopeServiceProvider is not available in production
- that's why the discoverer crashes because it cannot find the BaseTelescopeServiceProvider
- and the auto discoverer fails

Is it possible to just skip that class by catching the exception?

**Reply by ragingdave** [reply]
Yes you are correct there.
It is a class not found exception so potentially could catch it, but that feels a little invasive to basically just ignore classes where a parent class doesn't exist. This "hiding" might introduce a different equally hard to debug error in it's own right for certain situations.

**Reply by rubenvanassche** [reply]
Hmm I don't really see that as a problem, since we're only checking for classes that are a subclass of the `Settings` class, if there's a case where that class doesn't exist then something went really really wrong 😄 

Is the auto discoverer crashing here? https://github.com/spatie/laravel-settings/blob/37aa95acc0e0879b51244a504487773ca9822dba/src/Support/DiscoverSettings.php#L64

We could just add a try ... catch around that block and ignore it when it fails.

**Reply by ragingdave** [reply]
Yep that is exactly where it's failing. I did try a local mod with a try catch around that and it works but only if you are catching `Throwable`.

**Reply by rubenvanassche** [reply]
Added this functionality to 2.1.2, thanks for your input!
